src/relay/pass/ad.cc

```diff
@@ -0,0 +1,180 @@
+/*!
+ *  Copyright (c) 2018 by Contributors
+ * \file ad.cc
+ * \brief API for Automatic Differentiation for the Relay IR.
+ */
+
+#include <tvm/lowered_func.h>
+#include <tvm/operation.h>
+#include <tvm/relay/ad.h>
+#include <tvm/relay/builder.h>
+#include <tvm/relay/expr_functor.h>
+#include <tvm/relay/pass.h>
+#include "let_list.h"
+#include "../ir/type_functor.h"
+
+namespace tvm {
+namespace relay {
+
+using namespace tvm::runtime;
+
+Type WithGradientType(const Type& t) {
+  // TODO(M.K.): stricter checking
+  auto ty = t.as<FuncTypeNode>();
+  CHECK(ty) << "input should be a function";
+  return FuncTypeNode::make(ty->arg_types,
+                            TupleTypeNode::make({
+                              ty->ret_type,
+                              TupleTypeNode::make(ty->arg_types)}), {}, {});
+}
+
+//! \brief if the expression is a GlobalVar, transform to it's expression.
+Expr DeGlobal(const Module& mod, const Expr& e) {
+  if (auto x = e.as<GlobalVarNode>()) {
+    return mod->Lookup(GetRef<GlobalVar>(x))->body;
+  } else {
+    return e;
+  }
+}
+
+/*! \brief A fragment of the program being built by the automatic differentation
+ *  pass.
+ */
+struct ADValueNode {
+  virtual ~ADValueNode() { }
+  template <typename T>
+  T& get() {
+    auto ret = dynamic_cast<T*>(this);
+    CHECK(ret) << "cannot downcast";
+    return *ret;
+  }
+};
+
+using ADValue = std::shared_ptr<ADValueNode>;
+
+/*! \brief AD over a program which generates a tensor output. */
+struct ADTensor : ADValueNode {
+  Expr fwd;
+  mutable Expr rev;  // must be an id to avoid duplication
+  ADTensor(LetList* ll, const Expr& fwd) :
+    fwd(ll->Push(fwd)), rev(ll->Push(ZeroLike(this->fwd))) { }
+};
+
+/*! \brief A staged representation of the program, we reflect
+ * Relay functions into a function over fragments of AD. We
+ * can compute away this function to obtain a reverse mode program.
+ */
+struct ADFunction : ADValueNode {
+  std::function<ADValue(const std::vector<ADValue>&)> func;
+  explicit ADFunction(const std::function<ADValue(const std::vector<ADValue>&)>& func) :
+    func(func) { }
+};
+
+struct ReverseAD : ExprFunctor<ADValue(const Expr &)> {
+  const OpMap<FPrimalGradient> rev_map = Op::GetAttr<FPrimalGradient>("FPrimalGradient");
+  std::vector<std::function<void(LetList* ll)>> replay_actions;
```

Comment:
`actions` sounds strange, could you come up with a better name to indicate that the list is used to construct gradient expressions?